app/invocation/cache_requests_card.tsx

```diff
@@ -113,29 +151,162 @@ export default class CacheRequestsCardComponent extends React.Component<CacheReq
+    const value = Number(event.target.value) as cache.GetCacheScoreCardRequest.OrderBy;
+    // When changing the sort order, set direction according to a more useful
+    // default, to save an extra click. Asc is more useful for start time; Desc
+    // is more useful for duration.
+    const desc = value === cache.GetCacheScoreCardRequest.OrderBy.ORDER_BY_DURATION;
+    router.setQuery({
+      ...Object.fromEntries(this.props.search.entries()),
+      sort: String(value),
+      desc: String(desc),
+    });
+  }
+  private onToggleDescending() {
+    router.setQueryParam("desc", !this.getDescending());
+  }
+  private onChangeCacheType(event: React.ChangeEvent<HTMLSelectElement>) {
+    router.setQueryParam("cache", event.target.value);
+  }
+  private onChangeRequestType(event: React.ChangeEvent<HTMLSelectElement>) {
+    router.setQueryParam("request", event.target.value);
+  }
+  private onChangeResponseType(event: React.ChangeEvent<HTMLSelectElement>) {
+    router.setQueryParam("response", event.target.value);
+  }
+  private onToggleGroupByAction() {
+    router.setQueryParam("groupByAction", !this.getGroupByAction());
+  }
+  private searchTimeout = 0;
+  private onChangeSearch(event: React.ChangeEvent<HTMLInputElement>) {
+    const searchText = event.target.value;
+    // Update the search box state immediately, but debounce the query update so
+    // we don't fetch on each keystroke.
+    this.setState({ searchText });
+    clearTimeout(this.searchTimeout);
+    this.searchTimeout = window.setTimeout(
+      () => router.setQueryParam("search", searchText),
+      SEARCH_DEBOUNCE_INTERVAL_MS
+    );
+  }
+
   private onClickLoadMore() {
     this.fetchResults();
   }
 
+  private renderControls() {
+    return (
+      <>
+        <div className="controls row">
+          {/* Sorting controls */}
+          <label>Sort by</label>
+          <Select value={this.getOrderBy()} onChange={this.onChangeOrderBy.bind(this)}>
+            <Option value={cache.GetCacheScoreCardRequest.OrderBy.ORDER_BY_START_TIME}>Start time</Option>
+            <Option value={cache.GetCacheScoreCardRequest.OrderBy.ORDER_BY_DURATION}>Duration</Option>
+          </Select>
+          <OutlinedButton className="icon-button" onClick={this.onToggleDescending.bind(this)}>
+            {this.getDescending() ? <SortDesc className="icon" /> : <SortAsc className="icon" />}
+          </OutlinedButton>
+          {/* Filtering controls */}
+          <div className="separator" />
+          <Select value={this.getCacheType()} onChange={this.onChangeCacheType.bind(this)}>
+            <Option value={cache.CacheType.UNKNOWN_CACHE_TYPE}>CAS and AC</Option>
+            <Option value={cache.CacheType.CAS}>CAS only</Option>
+            <Option value={cache.CacheType.AC}>AC only</Option>
+          </Select>
+          <Select value={this.getRequestType()} onChange={this.onChangeRequestType.bind(this)}>
```

Comment:
People can see the same data that we show in the current UI by selecting cache="AC only", request="Read", response="NotFound". One thing we could do to make it easier is to make those the default filters.

That said, even if those are made the default settings, there is some room for improvement in how we present these filtering options to the user -- for example, to see all CAS transfers starting from those default filters, you'd have to change AC => CAS, Read => Read+Write, NotFound => All (a bit cumbersome IMO).

Maybe we could condense these 3 filters into a single "View" filter, with the options:
- All: cache=any, req=any, res=any
- Misses (default option): cache=AC, req=read, res=NotFound
- Hits: cache=AC, req=read, res=OK
- Downloads: cache=CAS, req=read, res=OK
- Uploads: cache=CAS, req=write, res=OK
- Errors: cache=any, req=any, res=(!NotFound && !OK)

But then I wonder whether people want to be able to see the AC requests and CAS requests side-by-side in the same view... the "All" option lets you do that, but then you don't get the option of filtering by Uploads vs. Downloads, or see only the CAS transfers involved in AC misses, etc.

Maybe a compromise between the 1-dropdown design (above) and 3-dropdown design (current), is to have two separate dropdowns, one for AC and one for CAS? Like:

AC (maybe labeled "Actions"?):
- All, Misses (default), Hits, Errors
CAS (maybe labeled "Artifacts"?):
- All, Hidden (default), Uploads, Downloads, Errors

Haven't thought this through a ton though -- would be interested to hear more about what you think

Reply:
I kinda like the single dropdown because that's how I think about it in my head (I feel like I understand this stuff pretty deeply, but I wasn't sure which options to select to see cache misses).

If people ask for more than those 6 views, we could always add new ones (or add advanced filtering options that gives them all 3).

Reply:
Combined into a single dropdown